pkg/pillar/cmd/zedrouter/networkinstance.go

```diff
@@ -1870,3 +1889,273 @@ func uplinkToPhysdev(ifname string) string {
+	if startIP == nil {
+		return types.AppNumInvalid, errors.New("parse error")
+	}
+	start := uint(startIP[0])<<24 + uint(startIP[1])<<16 +
+		uint(startIP[2])<<8 + uint(startIP[3])
+	ipAddr := staticIPAddr.To4()
+	addr := uint(ipAddr[0])<<24 + uint(ipAddr[1])<<16 +
+		uint(ipAddr[2])<<8 + uint(ipAddr[3])
+	appNumOnUNet = int(addr - start)
+	if appNumOnUNet > status.HostIPAddressCount {
+		appNumOnUNet = appNumOnUNet % status.HostIPAddressCount
+	}
+	// check whether this is already occupied
+	if status.AppNumBitMap.IsSet(appNumOnUNet) {
+		errStr := fmt.Sprintf("(%s-%s): IP Address (%s) is already occupied\n",
+			status.DisplayName, status.UUID.String(), ipAddr.String())
+		log.Errorf(errStr)
+		return types.AppNumInvalid, errors.New(errStr)
+	}
+	return appNumOnUNet, nil
+}
+
+// bridge to app number
+func uNetAppNumForBridgeGet(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus) (int, error) {
+	appNumOnUNet := types.AppNumInvalid
+	// a switch variant, ignore
+	if status.Type != types.NetworkInstanceTypeLocal &&
+		status.Type != types.NetworkInstanceTypeCloud {
+		return appNumOnUNet, nil
+	}
+	if status.HostIPAddressCount == 0 || status.Gateway == nil {
+		return appNumOnUNet, errors.New("no configuration")
+	}
+	appNumOnUNet, err := persistentAddrMapAppNumGet(ctx, status.UUID.String(),
+		status.BridgeName)
+	if err == nil {
+		return appNumOnUNet, nil
+	}
+	gateway := status.Gateway.To4()
+	if gateway == nil {
+		return 0, errors.New("gateway parse error")
+	}
+	startIP := status.DhcpRange.Start.To4()
+	if startIP == nil {
+		return appNumOnUNet, errors.New("dhcprange start address parse error")
+	}
+	start := uint(startIP[0])<<24 + uint(startIP[1])<<16 +
+		uint(startIP[2])<<8 + uint(startIP[3])
+	addr := uint(gateway[0])<<24 + uint(gateway[1])<<16 +
+		uint(gateway[2])<<8 + uint(gateway[3])
+	// if gateway is not within the DhcpRange
+	if addr < start {
+		appNumOnUNet = types.AppNumBridge
+		log.Functionf("uNetAppNumForBridgeGet(%s-%s): bridge AppNum %d\n",
+			status.DisplayName, status.UUID.String(), appNumOnUNet)
+		return appNumOnUNet, nil
+	}
+	appNumOnUNet = int(addr - start)
+	if appNumOnUNet > status.HostIPAddressCount {
+		appNumOnUNet = appNumOnUNet % status.HostIPAddressCount
+	}
+	if status.AppNumBitMap.IsSet(appNumOnUNet) {
+		return appNumOnUNet, errors.New("bridge IP is not available")
+	}
+	return appNumOnUNet, nil
+}
+
+// for an App Inastance attached to an underlay network,
+// try to get a free slot within the assignable ip address range
+func uNetAppNumAlloc(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus, appID string) (int, error) {
+	appNumOnUNet, err := persistentAddrMapAppNumGet(ctx,
+		status.UUID.String(), appID)
+	if err == nil {
+		return appNumOnUNet, nil
+	}
+	if status.HostIPAddressCount == 0 {
+		log.Functionf("uNetAppNumAlloc(%s-%s): not locally Addressable\n",
+			status.DisplayName, status.UUID.String())
+		return types.AppNumInvalid, nil
```

Comment:
I think we won't need this function as we split the appNumOnUNet and IP allocation apart, but I note that the caller doesn't check for AppNumInvalid. So it is unknow to me what happens as we hit this condition.

Reply:
AppNumInvalid is also for switch network bridges.  We are capturing the failures with Err.

Reply:
FWIW there is no code which ignores. So AppNumInvalid isn't needed at all.